Adding piped aggregations to working SQL query causes bad results or panic

## tl;dr

Starting from some working SQL extracted from mgbench bench3/q3, if I extend it with SuperSQL pipe syntax to perform an aggregation I either get incorrect data or an error.

## Details

Repro is with super commit 7beb4f9a and test data at `s3://brim-sampledata/mgbench/bench3.csup`.

In my work with benchmarks I've never been able to run the [original SQL bench3/q3 query](https://github.com/brimdata/mgbench-private/blob/main/bench3/q3/q3.sql) due to SuperSQL not yet supporting correlated subqueries and CTEs. I've also not been running its [SPQ equivalent](https://github.com/brimdata/mgbench-private/blob/main/bench3/q3/q3.spq) in vector runtime due to open issue https://github.com/brimdata/super/issues/5671. However that SPQ does run in sequential runtime and does produce results that match those from other systems that can run the original SQL query.

As a workaround to https://github.com/brimdata/super/issues/5671, I was following up on a suggestion from @nwt of just repeating the leading portion of the query into two inputs of one of our "regular" joins rather than using a self-join, but I quickly ran into a problem. To show it, we'll start with this working [SQL query fragment](https://github.com/brimdata/mgbench-private/blob/973961e35946acfaeac65650914c3c53da35b937/bench3/q3/q3.sql#L6-L35) that represents the contents of a CTE in the original SQL.

```
$ super -version
Version: v1.18.0-417-g7beb4f9a1

$ SUPER_VAM=1 super -z -c "
  SELECT dow,
         hr,
         device_name,
         AVG(motions) AS in_use
  FROM (
    SELECT dt,
           dow,
           hr,
           device_name,
           COUNT(*) AS motions
    FROM (
      SELECT CAST(log_time AS DATE) AS dt,
             EXTRACT(DOW FROM log_time) AS dow,
             EXTRACT(HOUR FROM log_time) AS hr,
             device_name
      FROM 'bench3.csup'
      WHERE device_name LIKE 'room%'
        AND event_type = 'motion_start'
        AND log_time >= TIMESTAMP '2019-09-01 00:00:00'
    ) AS r
    WHERE dow IN (1,2,3,4,5)
      AND hr BETWEEN 9 AND 16
    GROUP BY dt,
             dow,
             hr,
             device_name
  ) AS s
  GROUP BY dow,
           hr,
           device_name"

{dow:3,hr:15,device_name:"room_169",in_use:15.}
{dow:1,hr:16,device_name:"room_506",in_use:5.333333333333333}
{dow:3,hr:10,device_name:"room_167",in_use:1.5}
...
```

However, if I extend it with a pipe and an aggregation like I use [in one of the forks leading into the self-join in my SPQ variation](https://github.com/brimdata/mgbench-private/blob/973961e35946acfaeac65650914c3c53da35b937/bench3/q3/q3.spq#L41-L43), now I'm shown strange field names like `k2` and `t0` that weren't in the output when I ran the SQL fragment.

```
$ SUPER_VAM=1 super -z -c "
  SELECT dow,
         hr,
         device_name,
         AVG(motions) AS in_use
  FROM (
    SELECT dt,
           dow,
           hr,
           device_name,
           COUNT(*) AS motions
    FROM (
      SELECT CAST(log_time AS DATE) AS dt,
             EXTRACT(DOW FROM log_time) AS dow,
             EXTRACT(HOUR FROM log_time) AS hr,
             device_name
      FROM 'bench3.csup'
      WHERE device_name LIKE 'room%'
        AND event_type = 'motion_start'
        AND log_time >= TIMESTAMP '2019-09-01 00:00:00'
    ) AS r
    WHERE dow IN (1,2,3,4,5)
      AND hr BETWEEN 9 AND 16
    GROUP BY dt,
             dow,
             hr,
             device_name
  ) AS s
  GROUP BY dow,
           hr,
           device_name
| aggregate
       in_use := min(in_use)
       by device_name"

{k2:"room_201",t0:1.}
{k2:"room_271",t0:1.6666666666666667}
{k2:"room_506",t0:1.}
{k2:"room_316",t0:3.}
{k2:"room_169",t0:1.}
{k2:"room_167",t0:1.}
{k2:"room_205",t0:1.5}
{k2:"room_410",t0:1.}
{k2:"room_477",t0:1.}
{k2:"room_227",t0:1.}
{k2:"room_348",t0:1.}
{k2:"room_265",t0:1.}
```

Puzzled by this, I instead did a different aggregation and that gave me a panic.

```
$ SUPER_VAM=1 super -z -c "
  SELECT dow,
         hr,
         device_name,
         AVG(motions) AS in_use
  FROM (
    SELECT dt,
           dow,
           hr,
           device_name,
           COUNT(*) AS motions
    FROM (
      SELECT CAST(log_time AS DATE) AS dt,
             EXTRACT(DOW FROM log_time) AS dow,
             EXTRACT(HOUR FROM log_time) AS hr,
             device_name
      FROM 'bench3.csup'
      WHERE device_name LIKE 'room%'
        AND event_type = 'motion_start'
        AND log_time >= TIMESTAMP '2019-09-01 00:00:00'
    ) AS r
    WHERE dow IN (1,2,3,4,5)
      AND hr BETWEEN 9 AND 16
    GROUP BY dt,
             dow,
             hr,
             device_name
  ) AS s
  GROUP BY dow,
           hr,
           device_name
| count() by device_name"

panic: interface conversion: dag.Expr is *dag.Dot, not *dag.This

goroutine 1 [running]:
github.com/brimdata/super/compiler/kernel.(*Builder).compileVamAggregate(0xc0000f7810, 0xc00078ed80, {0x8f78560, 0xc000487550})
	/Users/phil/work/super/compiler/kernel/vop.go:412 +0x85d
github.com/brimdata/super/compiler/kernel.(*Builder).compileVamLeaf(0xc0000f7810, {0x8f767a0?, 0xc00078ed80}, {0x8f78560, 0xc000487550})
	/Users/phil/work/super/compiler/kernel/vop.go:215 +0x97
github.com/brimdata/super/compiler/kernel.(*Builder).compileVam(0xc0000f7810, {0x8f767a0?, 0xc00078ed80}, {0xc00069df40, 0x1, 0xc00088bd80?})
	/Users/phil/work/super/compiler/kernel/vop.go:85 +0x3f7
github.com/brimdata/super/compiler/kernel.(*Builder).compileVamSeq(0xc0000f7810, {0xc00088b780?, 0x0?, 0xc00064d788?}, {0x0?, 0xc00088bd80?, 0x4?})
	/Users/phil/work/super/compiler/kernel/vop.go:398 +0x6f
github.com/brimdata/super/compiler/kernel.(*Builder).Build(0xc0000f7810, {0xc00088b780, 0x8, 0x8}, {0x0, 0x0, 0x0})
	/Users/phil/work/super/compiler/kernel/op.go:91 +0xed
github.com/brimdata/super/compiler.Build(0xc0005367b0, {0xc00088b780, 0x8, 0x8}, 0xc0003e8800, {0x0, 0x0, 0x0})
	/Users/phil/work/super/compiler/package.go:60 +0x1e5
github.com/brimdata/super/compiler.CompileWithAST(0xc0005367b0, 0x716c6a5?, 0xc0003e8800, 0x1, 0xc, {0x0, 0x0, 0x0})
	/Users/phil/work/super/compiler/package.go:87 +0xe5
github.com/brimdata/super/compiler.(*compiler).NewQuery(0x8f912a8?, 0xc000128820?, 0xc00064d9e0?, {0x0?, 0x8?, 0x8d45460?}, 0x7095f00?)
	/Users/phil/work/super/compiler/compiler.go:38 +0x45
github.com/brimdata/super/runtime.CompileQuery({0x8f912a8?, 0xc000128820?}, 0xc00013be60, {0x8f7f170, 0xc000114140}, 0xc0003e8760, {0x0, 0x0, 0x0})
	/Users/phil/work/super/runtime/compiler.go:34 +0x10f
github.com/brimdata/super/cmd/super/root.(*Command).Run(0xc00016a200, {0xc00003c0f0, 0x0, 0x0})
	/Users/phil/work/super/cmd/super/root/command.go:158 +0x45d
github.com/brimdata/super/pkg/charm.path.run({0xc000114250, 0x1, 0x1}, {0xc00003c0f0, 0x0, 0x0?})
	/Users/phil/work/super/pkg/charm/path.go:11 +0x7b
github.com/brimdata/super/pkg/charm.(*Spec).Exec(0xa0c5200, {0xc00003c0d0, 0x3, 0x3})
	/Users/phil/work/super/pkg/charm/charm.go:71 +0x105
main.main()
	/Users/phil/work/super/cmd/super/main.go:40 +0x5b
```

Of course, if I save off the result from the SQL fragment and run the aggregations on that as a separate step, it works fine.

```
$ SUPER_VAM=1 super -f csup -c "
  SELECT dow,
         hr,
         device_name,
         AVG(motions) AS in_use
  FROM (
    SELECT dt,
           dow,
           hr,
           device_name,
           COUNT(*) AS motions
    FROM (
      SELECT CAST(log_time AS DATE) AS dt,
             EXTRACT(DOW FROM log_time) AS dow,
             EXTRACT(HOUR FROM log_time) AS hr,
             device_name
      FROM 'bench3.csup'
      WHERE device_name LIKE 'room%'
        AND event_type = 'motion_start'
        AND log_time >= TIMESTAMP '2019-09-01 00:00:00'
    ) AS r
    WHERE dow IN (1,2,3,4,5)
      AND hr BETWEEN 9 AND 16
    GROUP BY dt,
             dow,
             hr,
             device_name
  ) AS s
  GROUP BY dow,
           hr,
           device_name" > pre-agg.csup

$ SUPER_VAM=1 super -z -c "
from 'pre-agg.csup'
| aggregate
       in_use := min(in_use)
       by device_name"
{device_name:"room_205",in_use:1.}
{device_name:"room_169",in_use:1.}
{device_name:"room_227",in_use:1.}
{device_name:"room_477",in_use:1.}
{device_name:"room_265",in_use:1.}
{device_name:"room_316",in_use:1.}
{device_name:"room_167",in_use:1.}
{device_name:"room_410",in_use:1.}
{device_name:"room_348",in_use:1.}
{device_name:"room_506",in_use:1.}
{device_name:"room_271",in_use:1.}
{device_name:"room_201",in_use:1.}

$ SUPER_VAM=1 super -z -c "
from 'pre-agg.csup'
| count() by device_name"
{device_name:"room_410",count:37(uint64)}
{device_name:"room_506",count:26(uint64)}
{device_name:"room_205",count:21(uint64)}
{device_name:"room_271",count:36(uint64)}
{device_name:"room_227",count:36(uint64)}
{device_name:"room_348",count:19(uint64)}
{device_name:"room_201",count:37(uint64)}
{device_name:"room_477",count:38(uint64)}
{device_name:"room_167",count:40(uint64)}
{device_name:"room_169",count:18(uint64)}
{device_name:"room_265",count:10(uint64)}
{device_name:"room_316",count:37(uint64)}
```

--- a/compiler/optimizer/optimizer.go
+++ b/compiler/optimizer/optimizer.go
@@ -149,7 +149,7 @@ func (o *Optimizer) Optimize(seq dag.Seq) (dag.Seq, error) {
 	seq = liftFilterOps(seq)
 	seq = mergeFilters(seq)
 	seq = mergeYieldOps(seq)
-	seq = inlineRecordExprSpreads(seq)
+	inlineRecordExprSpreads(seq)
 	seq = removePassOps(seq)
 	seq = replaceSortAndHeadOrTailWithTop(seq)
 	o.optimizeParallels(seq)
@@ -496,8 +496,8 @@ func matchFilter(seq dag.Seq) (dag.Expr, dag.Seq) {
 }
 
 // inlineRecordExprSpreads transforms "{...{a}}" to "{a}".
-func inlineRecordExprSpreads(seq dag.Seq) dag.Seq {
-	walkT(reflect.ValueOf(seq), func(r *dag.RecordExpr) *dag.RecordExpr {
+func inlineRecordExprSpreads(v any) {
+	walkT(reflect.ValueOf(v), func(r *dag.RecordExpr) *dag.RecordExpr {
 		for i := range r.Elems {
 			s, ok := r.Elems[i].(*dag.Spread)
 			if !ok {
@@ -511,7 +511,6 @@ func inlineRecordExprSpreads(seq dag.Seq) dag.Seq {
 		}
 		return r
 	})
-	return seq
 }
 
 func liftFilterOps(seq dag.Seq) dag.Seq {
@@ -566,25 +565,34 @@ func mergeYieldOps(seq dag.Seq) dag.Seq {
 			if !ok {
 				continue
 			}
-			switch seq[i+1].(type) {
-			case *dag.Aggregate, *dag.Yield:
+			propagateY1Fields := func(e dag.Expr) dag.Expr {
+				this, ok := e.(*dag.This)
+				if !ok {
+					return e
+				}
+				y1Expr, ok := y1TopLevelFields[this.Path[0]]
+				if !ok {
+					if y1TopLevelSpread != nil {
+						return addPathToExpr(y1TopLevelSpread, this.Path)
+					}
+					return e
+				}
+				return addPathToExpr(y1Expr, this.Path[1:])
+			}
+			switch op := seq[i+1].(type) {
+			case *dag.Aggregate:
+				for i := range op.Keys {
+					walkT(reflect.ValueOf(&op.Keys[i].RHS), propagateY1Fields)
+				}
+				for i := range op.Aggs {
+					walkT(reflect.ValueOf(&op.Aggs[i].RHS), propagateY1Fields)
+				}
+			case *dag.Yield:
+				walkT(reflect.ValueOf(op.Exprs), propagateY1Fields)
 			default:
 				continue
 			}
-			walkT(reflect.ValueOf(seq[i+1]), func(e2 dag.Expr) dag.Expr {
-				this2, ok := e2.(*dag.This)
-				if !ok {
-					return e2
-				}
-				e1, ok := y1TopLevelFields[this2.Path[0]]
-				if !ok {
-					if y1TopLevelSpread != nil {
-						return addPathToExpr(y1TopLevelSpread, this2.Path)
-					}
-					return e2
-				}
-				return addPathToExpr(e1, this2.Path[1:])
-			})
+			inlineRecordExprSpreads(seq[i+1])
 			seq.Delete(i, i+1)
 			i--
 		}

--- a/compiler/ztests/merge-yields.yaml
+++ b/compiler/ztests/merge-yields.yaml
@@ -10,6 +10,8 @@ script: |
   super compile -C -O '{...a,...b} | yield {c}'
   echo ===
   super compile -C -O '{a,...b} | yield {a}'
+  echo ===
+  super compile -C -O 'select a, max(b) as c group by a | aggregate min(c) by a'
 
 outputs:
   - name: stdout
@@ -39,4 +41,11 @@ outputs:
       null
       | yield {a:a,...b}
       | yield {a:a}
+      | output main
+      ===
+      null
+      | aggregate
+          t0:=max(b) by k0:=a
+      | aggregate
+          min:=min(t0) by a:=k0
       | output main

